# REFER TO "CS AND MATH DATAFRAME" FOR BEGINNING OF THIS NOTEBOOK UP TO NEXT COMMENT

# In this, quick analysis is done on the classes that only have one class and in turn one professor, so no math has to be done at the moment. Finishes off with a dataframe that has all the CS and Math courses left after taking care of one off classes

In [2]:
import pandas as pd
import numpy as np
url = "https://docs.google.com/spreadsheets/d/1mS6khEB6m8cPNenNvY9Tg6bJ6YkmcvCI/export?format=csv&gid=1283335856"
df = pd.read_csv(url)

In [3]:
df.head()

TERM SUBJECT  NBR                  COURSE NAME SECTION          PROF  \
0  F2022    ACCT  100               Fin & Mgr Acct       1         HO, V   
1  F2022    ACCT  101  Intro Thry & Prac of Acct I      11       CHAN, J   
2  F2022    ACCT  101  Intro Thry & Prac of Acct I       5   RUTHIZER, S   
3  F2022    ACCT  101  Intro Thry & Prac of Acct I       2      GRUZA, M   
4  F2022    ACCT  101  Intro Thry & Prac of Acct I       4  FEISULLIN, A   

   TOTAL  A+   A  A-  ...  C  C-  D  F   W  INC/NA  AVG GPA  \
0     28   0  11   2  ...  3   0  0  0   5       0    3.383   
1     29   0  13   2  ...  0   1  0  0   3       0    3.464   
2     50   2   9  14  ...  1   0  3  2   3       0    3.176   
3     45   1   2   3  ...  0   0  0  0  25       0    3.067   
4     45   4   6   1  ...  9   1  1  0   4       1    2.908   

   join the best club on campus  Unnamed: 22  https://discord.gg/vxCqd4hdak  
0                           NaN          NaN                            NaN  
1                           NaN          NaN                            NaN  
2                           NaN          NaN                            NaN  
3                           NaN          NaN                            NaN  
4                           NaN          NaN                            NaN  

[5 rows x 24 columns]

#### I include an extra column "COURSE NAME" to make it easier to know which class number is related to which class name, and though I may know the class names for these specific departments fairly well, there will be other departments I may am unfamilar with

#### I also implement a function that converts a gpa into a letter grade which will be used throughout the analysis of all subjects

In [4]:
important_columns = ["SUBJECT", "NBR", "COURSE NAME", "PROF", "TOTAL", "A+", "A", "A-", "B+", "B", "B-", "C+", "C", "C-", "D", "F", "W", "AVG GPA"]
analysis_df = df[important_columns]

In [5]:
analysis_df.head()

SUBJECT  NBR                  COURSE NAME          PROF  TOTAL  A+   A  A-  \
0    ACCT  100               Fin & Mgr Acct         HO, V     28   0  11   2   
1    ACCT  101  Intro Thry & Prac of Acct I       CHAN, J     29   0  13   2   
2    ACCT  101  Intro Thry & Prac of Acct I   RUTHIZER, S     50   2   9  14   
3    ACCT  101  Intro Thry & Prac of Acct I      GRUZA, M     45   1   2   3   
4    ACCT  101  Intro Thry & Prac of Acct I  FEISULLIN, A     45   4   6   1   

   B+  B  B-  C+  C  C-  D  F   W  AVG GPA  
0   1  3   3   0  3   0  0  0   5    3.383  
1   0  4   5   0  0   1  0  0   3    3.464  
2   6  4   5   0  1   0  3  2   3    3.176  
3   1  5   0   6  0   0  0  0  25    3.067  
4   7  5   0   6  9   1  1  0   4    2.908

In [6]:
analysis_df = analysis_df.dropna(subset = "PROF")

In [7]:
analysis_df.shape

(2537, 18)

In [8]:
analysis_df = analysis_df[analysis_df["AVG GPA"] != 0]
analysis_df.shape

(2333, 18)

In [9]:
CS_df = analysis_df.loc[analysis_df["SUBJECT"] == "CSCI"]
MATH_df = analysis_df.loc[analysis_df["SUBJECT"] == "MATH"]
CS_MATH_df = pd.concat([CS_df, MATH_df])
CS_MATH_df.shape

(245, 18)

#### This allows me to display all the rows in my dataframe by default
#### Reset this back by doing "pd.reset_option('display.max_rows')"

In [10]:
pd.set_option('display.max_rows', None)
CS_MATH_df

SUBJECT   NBR                     COURSE NAME               PROF  TOTAL  \
582     CSCI    12   Intro Computers & Computation          WAXMAN, J     58   
585     CSCI    48         Spreadsheet Programming           FRIED, M    131   
586     CSCI    48         Spreadsheet Programming          JAGDEO, M     94   
596     CSCI   100    Information and Intelligence          CASTRO, R     40   
597     CSCI   111  Intro Algorithmic Problem Solv          SCHLEY, R    236   
598     CSCI   111  Intro Algorithmic Problem Solv            CHYN, X    239   
599     CSCI   111  Intro Algorithmic Problem Solv            CHYN, X    240   
619     CSCI   112  Intro Algori Prob Solving Java          CONNOR, T     13   
621     CSCI   211  Object-Oriented Program in C++          WAXMAN, J    150   
622     CSCI   211  Object-Oriented Program in C++          WAXMAN, J    105   
623     CSCI   211  Object-Oriented Program in C++          ALAYEV, Y     16   
635     CSCI   212    Object-Oriented Prog in Java             TSE, C     36   
636     CSCI   212    Object-Oriented Prog in Java            LORD, K    161   
637     CSCI   212    Object-Oriented Prog in Java            LORD, K    127   
648     CSCI   220             Discrete Structures       KAHROBAEI, D     39   
649     CSCI   220             Discrete Structures       KAHROBAEI, D     42   
650     CSCI   220             Discrete Structures            CHYN, X     88   
651     CSCI   220             Discrete Structures            CHYN, X     58   
652     CSCI   220             Discrete Structures       CESARETTI, P     30   
653     CSCI   240    Computer Org & Assembly Lang             YEH, J     45   
654     CSCI   240    Computer Org & Assembly Lang             YEH, J     45   
655     CSCI   240    Computer Org & Assembly Lang             YEH, J     45   
656     CSCI   240    Computer Org & Assembly Lang          SVITAK, J     31   
657     CSCI   240    Computer Org & Assembly Lang          SVITAK, J     26   
658     CSCI   240    Computer Org & Assembly Lang          SVITAK, J     37   
659     CSCI   313                 Data Structures             TAO, X     39   
660     CSCI   313                 Data Structures             TAO, X     36   
661     CSCI   313                 Data Structures             TAO, X     24   
662     CSCI   313                 Data Structures            TSAI, C     10   
663     CSCI   313                 Data Structures       STEINBERG, O     31   
664     CSCI   313                 Data Structures  SMITH-THOMPSON, A     29   
665     CSCI   313                 Data Structures       STEINBERG, O     42   
666     CSCI   313                 Data Structures  SMITH-THOMPSON, A     25   
667     CSCI   316  Principles of Programming Lang  SMITH-THOMPSON, A     35   
668     CSCI   316  Principles of Programming Lang            KONG, T     35   
669     CSCI   316  Principles of Programming Lang            KONG, T     35   
670     CSCI   316  Principles of Programming Lang          SVITAK, J     28   
671     CSCI   316  Principles of Programming Lang          SVITAK, J     30   
672     CSCI   316  Principles of Programming Lang            KONG, T     61   
673     CSCI   320           Theory of Computation         OBRENIC, B     86   
674     CSCI   320           Theory of Computation          BOKLAN, K     37   
675     CSCI   320           Theory of Computation         OBRENIC, B    109   
676     CSCI   320           Theory of Computation          BOKLAN, K     28   
677     CSCI   323    Design & Analysis Algorithms             TAO, X     30   
678     CSCI   323    Design & Analysis Algorithms        PHILLIPS, T     18   
679     CSCI   323    Design & Analysis Algorithms        PHILLIPS, T     25   
680     CSCI   323    Design & Analysis Algorithms           BROWN, T     30   
681     CSCI   323    Design & Analysis Algorithms         GOSWAMI, M     61   
682     CSCI   331                Database Systems       STEINBERG, O     35   
683  

# I'll first record all the classes in the CS and Math department that only have one class offered 

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.index.html

In [11]:
value_counts = CS_MATH_df["NBR"].value_counts() 
unique_values = value_counts[value_counts == 1].index
one_class_df = CS_MATH_df[CS_MATH_df['NBR'].isin(unique_values)]

one_class_df   

SUBJECT   NBR                     COURSE NAME            PROF  TOTAL  A+  \
582     CSCI    12   Intro Computers & Computation       WAXMAN, J     58   0   
596     CSCI   100    Information and Intelligence       CASTRO, R     40   0   
619     CSCI   112  Intro Algori Prob Solving Java       CONNOR, T     13   6   
687     CSCI   334     Data Mining and Warehousing           SY, B     13   0   
688     CSCI   335   Information Org and Retrieval     GOLDBERG, R     33   0   
703     CSCI   344             Distributed Systems      FLUTURE, S     22   1   
704     CSCI   348             Data Communications       RAHMAN, M     26   0   
1762    MATH   114            Elementary Prob&Stat       LEHMAN, S     20   1   
1829    MATH   132            Calc Appl Soc Sci II      GOLDMAN, S     12   1   
1872    MATH   209           Elementary Set Theory       KLOSIN, K     14   0   
1897    MATH   242      Methods Of Math Statistics       SISSER, F     36   7   
1898    MATH   247             Lin Prog & Game Thy       KLOSIN, K     36   4   
1899    MATH   301              Abstract Algebra I       MILLER, R     15   0   
1900    MATH   310        Elementary Real Analysis       WILSON, S     24   0   
1901    MATH   505    Mathematical Problem-Solving     GANGARAM, E     30   1   
1902    MATH   509            Set Theory and Logic       KLOSIN, K     19   0   
1903    MATH   518                College Geometry    MARKINSON, M     21   2   
1904    MATH   614     Functions of Real Variables        MITRA, S     11   2   
1905    MATH   621                     Probability      RALESCU, S     24   0   
1906    MATH   624           Numerical Analysis I.  OVCHINNIKOV, A     15   3   
1907    MATH   634                Theory of Graphs       HANUSA, C     12   5   
1908    MATH  2902          Studies in Mathematics      ZOMBORY, B     10   1   

       A  A-  B+   B  B-  C+  C  C-  D  F   W  AVG GPA  
582    2   7   5  14   6   2  2   1  3  5  10    2.594  
596    8   4   5   5   3   0  4   2  0  3   6    2.876  
619    2   1   1   1   1   0  0   0  0  0   1    3.725  
687    1   2   0   2   0   1  1   0  0  0   6    3.100  
688    4   4   4   5   3   2  5   0  0  1   5    2.918  
703    1   0   0   1   2   2  3   0  0  1  10    2.455  
704    2   5   3   6   3   1  4   0  0  1   1    2.912  
1762   1   4   0   5   0   2  0   1  3  0   3    2.771  
1829   0   1   2   2   1   1  0   0  1  0   2    2.922  
1872   2   1   0   0   1   4  0   0  3  1   2    2.217  
1897   2   1   0   0   4   0  3   4  0  8   7    2.183  
1898  16   0   1   1   4   2  2   0  0  2   4    3.303  
1899   0   0   0   0   0   0  0   1  0  1  13    0.850  
1900   5   3   0   4   2   2  2   1  0  0   5    3.095  
1901   2   3   2   4   5   1  5   3  0  4   0    2.420  
1902   3   0   0   2   3   2  1   0  0  5   3    2.044  
1903   5   1   5   1   1   1  1   0  0  3   1    2.910  
1904   3   2   2   1   1   0  0   0  0  0   0    3.609  
1905   3   2   1   3   1   0  1   1  0  2  10    2.721  
1906   3   2   0   1   0   1  1   0  0  1   3    3.225  
1907   4   0   0   0   0   0  0   1  0  1   1    3.427  
1908   3   2   0   1   0   0  0   0  0  0   2    3.771

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html

In [12]:
CS_MATH_df.dtypes

SUBJECT         object
NBR             object
COURSE NAME     object
PROF            object
TOTAL            int64
A+               int64
A                int64
A-               int64
B+               int64
B                int64
B-               int64
C+               int64
C                int64
C-               int64
D                int64
F                int64
W                int64
AVG GPA        float64
dtype: object

#### Makes the values in the column "NBR" floats so that I can take out classes where the class number is 399 or less. I also take into account the one math class that can't be changed into a numeric, which I probably won't use: when dealing with class numbers that have letters in them, I will have to deal with them before I change that column to hold only numeric values. This also means I can't make all "NBR" values numeric at the very beginning of bringing in any semester's worth of data

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html

In [13]:
CS_MATH_df['NBR'] = pd.to_numeric(CS_MATH_df['NBR'], errors='coerce')
math_class_name_with_letter = CS_MATH_df['NBR'].fillna("114W")

CS_MATH_df.dtypes

SUBJECT         object
NBR            float64
COURSE NAME     object
PROF            object
TOTAL            int64
A+               int64
A                int64
A-               int64
B+               int64
B                int64
B-               int64
C+               int64
C                int64
C-               int64
D                int64
F                int64
W                int64
AVG GPA        float64
dtype: object

In [15]:
print(math_class_name_with_letter)

582       12.0
585       48.0
586       48.0
596      100.0
597      111.0
598      111.0
599      111.0
619      112.0
621      211.0
622      211.0
623      211.0
635      212.0
636      212.0
637      212.0
648      220.0
649      220.0
650      220.0
651      220.0
652      220.0
653      240.0
654      240.0
655      240.0
656      240.0
657      240.0
658      240.0
659      313.0
660      313.0
661      313.0
662      313.0
663      313.0
664      313.0
665      313.0
666      313.0
667      316.0
668      316.0
669      316.0
670      316.0
671      316.0
672      316.0
673      320.0
674      320.0
675      320.0
676      320.0
677      323.0
678      323.0
679      323.0
680      323.0
681      323.0
682      331.0
683      331.0
684      331.0
685      331.0
686      331.0
687      334.0
688      335.0
689      340.0
690      340.0
691      340.0
692      340.0
693      340.0
694      340.0
695      340.0
696      343.0
697      343.0
698      343.0
699      343.0
700      3

#### Getting an updated dataframe of one off classes in the CS or Math department which now stay with only normal undergraduate classes

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.index.html

In [16]:
value_counts = CS_MATH_df["NBR"].value_counts() 
unique_values = value_counts[value_counts == 1].index 
one_class_df = CS_MATH_df[(CS_MATH_df['NBR'].isin(unique_values)) & (CS_MATH_df['NBR'] <=399)]

one_class_df 

SUBJECT    NBR                     COURSE NAME         PROF  TOTAL  A+  \
582     CSCI   12.0   Intro Computers & Computation    WAXMAN, J     58   0   
596     CSCI  100.0    Information and Intelligence    CASTRO, R     40   0   
619     CSCI  112.0  Intro Algori Prob Solving Java    CONNOR, T     13   6   
687     CSCI  334.0     Data Mining and Warehousing        SY, B     13   0   
688     CSCI  335.0   Information Org and Retrieval  GOLDBERG, R     33   0   
703     CSCI  344.0             Distributed Systems   FLUTURE, S     22   1   
704     CSCI  348.0             Data Communications    RAHMAN, M     26   0   
1762    MATH  114.0            Elementary Prob&Stat    LEHMAN, S     20   1   
1829    MATH  132.0            Calc Appl Soc Sci II   GOLDMAN, S     12   1   
1872    MATH  209.0           Elementary Set Theory    KLOSIN, K     14   0   
1897    MATH  242.0      Methods Of Math Statistics    SISSER, F     36   7   
1898    MATH  247.0             Lin Prog & Game Thy    KLOSIN, K     36   4   
1899    MATH  301.0              Abstract Algebra I    MILLER, R     15   0   
1900    MATH  310.0        Elementary Real Analysis    WILSON, S     24   0   

       A  A-  B+   B  B-  C+  C  C-  D  F   W  AVG GPA  
582    2   7   5  14   6   2  2   1  3  5  10    2.594  
596    8   4   5   5   3   0  4   2  0  3   6    2.876  
619    2   1   1   1   1   0  0   0  0  0   1    3.725  
687    1   2   0   2   0   1  1   0  0  0   6    3.100  
688    4   4   4   5   3   2  5   0  0  1   5    2.918  
703    1   0   0   1   2   2  3   0  0  1  10    2.455  
704    2   5   3   6   3   1  4   0  0  1   1    2.912  
1762   1   4   0   5   0   2  0   1  3  0   3    2.771  
1829   0   1   2   2   1   1  0   0  1  0   2    2.922  
1872   2   1   0   0   1   4  0   0  3  1   2    2.217  
1897   2   1   0   0   4   0  3   4  0  8   7    2.183  
1898  16   0   1   1   4   2  2   0  0  2   4    3.303  
1899   0   0   0   0   0   0  0   1  0  1  13    0.850  
1900   5   3   0   4   2   2  2   1  0  0   5    3.095

#### I wanted to make a code using pandas that does the same thing as the google sheets does for this grade distribution online

https://www.w3schools.com/python/python_lambda.asp

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html

In [17]:

# Define the mapping of letter grades to numeric values
grade_mapping = {
    'A+': 4.0,
    'A': 4.0,
    'A-': 3.7,
    'B+': 3.3,
    'B': 3.0,
    'B-': 2.7,
    'C+': 2.3,
    'C': 2.0,
    'C-': 1.7,
    'D': 1.0,
    'F': 0.0
}

grades = ["A+", "A", "A-", "B+", "B", "B-", "C+", "C", "C-", "D", "F"]

# Calculate the weighted average grade for each class
weighted_avg_grades = one_class_df[grades].apply(lambda row: sum(row[column] * grade_mapping[column] for column in one_class_df[grades].columns) / row.sum(), axis=1)

# Print the average letter grades for each class
print(weighted_avg_grades)

582     2.593617
596     2.876471
619     3.725000
687     3.100000
688     2.917857
703     2.454545
704     2.912000
1762    2.770588
1829    2.922222
1872    2.216667
1897    2.182759
1898    3.303125
1899    0.850000
1900    3.094737
dtype: float64


## CSCI 12 (Intro Computers and Computation) - Concepts of thinking computationally and the technical tools to function effectively in the digitally-connected world. Hands-on introduction to computers, computation, the Internet, and the basics of computer hardware and software. Students will have experience during the instructed computer lab with a number of software environments including an operating system, a word processor, a presentation tool, a spreadsheet, and a database system. The course will focus on problem solving and programming within the context of a wide range of applications, and prepare students for further study in a variety of majors and minors. In addition, students will acquire the skills needed to learn other similar tools on their own. Two and one-half years of high school mathematics is required 

### Professor: Jerry Waxman, AVG GPA: 2.6 (https://www.ratemyprofessors.com/professor/287312) which is a C+

## CSCI 100 (Information and Intelligence) - How information measurement, encoding, and transmission relate to the design of artificial intelligence agents such as search engines, robots, and programs that mimic human intelligence. Models of human and artificial intelligence; relations among information, meaning, and data; diagnostic and causal reasoning in the presence of uncertainty. Readings from the literature of information theory and artificial intelligence; writing assignments, completion of a project to design and/or construct an information-driven intelligent agent 

### Professor: Renne Castro, AVG GPA: 2.9 (https://www.ratemyprofessors.com/professor/2368930) which is a B-

## CSCI 112 (Intro Algorithmic Problem Solving Java) - Introduction to the principles and practice of programming. Topics include primitive data types; concepts of object, class, and method; control structures; arrays procedures and functions; parameter passing; scope and lifetime of variables; input and output; documentation. 

### Professor: Timonthy Connor, AVG GPA: 3.7 (https://www.ratemyprofessors.com/professor/2455397) which is a A

## CSCI 334 (Data Mining and Warehousing) - Data Mining and data warehousing: data warehouse basics; concept of patterns and visualization; information theory, information and statistics linkage; temporal-spatial data; change point detection; statistical association patterns; pattern inference and model discovery; Bayesian networks; pattern ordering inference; selected case study 

### Professor: Bon Sy, AVG GPA: 3.1, however, half the class withdrew (https://www.ratemyprofessors.com/professor/217852) which is a B

## CSCI 335 (Information Organization and Retrieval) - Concepts of information retrieval: keywords and Boolean retrieval; text processing, automatic indexing, term weighting, similarity measures; retrieval models: vector model, probabilistic model; extended Boolean systems: fuzzy set, p-norm models; linguistic model; extensions and AI techniques: learning and relevance feedback; term dependence; document and term clustering;network approaches; linguistic analysis and knowledge representation. Implementation: inverted files; efficiency issues for large-scale systems; integrating database and information retrieval

### Professor: Robert Goldberg, AVG GPA: 2.9 (https://www.ratemyprofessors.com/professor/446485) which is a B-

## CSCI 344 (Distributed Systems) - Issues in the implementation of computer systems using multiple processors linked through a communication network. Communication in distributed systems including layered protocols and the client-server model; synchronization of distributed processes and process threads

### Professor: Simina Fluture, AVG GPA: 2.5, and half the class dropped (https://www.ratemyprofessors.com/professor/513427) which is a C+

## CSCI 348 (Data Communications) - Computer communications and networks; carriers, media, interfaces (RS 232, RS 422, CCITT); circuit types, data codes, synchronous and asynchronous transmission; protocols (OSI, TCP/IP); modems, multiplexors, and other network hardware; error correction and encryption; voice and data switching: local area networks, ISDN, packet switching; issues in the architecture, design, and management of networks 

### Professor: Md Mahbubur Rahman, AVG GPA: 2.9 (https://www.ratemyprofessors.com/professor/2649371) which is a B-

## Math 114 (Elementary Prob and Stats) - An introduction to mathematical probability and statistics for the general student with a writing-intensive component. Includes the material in MATH 114, as well as additional topics such as sampling methods, research design, and composing and conducting surveys, explored through student research and writing assignments. Not open to mathematics, physics, or chemistry majors, or to students who are taking or have passed MATH 114, 241, 611, 621, 633, BIOL 230, ECON 249, PSYCH 107, SOC 205, 206, 207. Not open to students who will be receiving transfer credit or advanced placement credit for MATH 114 

### Professor: Seth Lehman, AVG GPA: 2.8 (https://www.ratemyprofessors.com/professor/2447352) which is a B-

## Math 132 (Calculus Applied Social Science II) - A continuation of MATH 131. Topics include limits and continuity; mean value theorem; antiderivatives; integrals and integration techniques; applications of the definite integral; the calculus of logarithmic, exponential, and trigonometric functions. This course prepares students who have taken MATH 131 to continue into MATH 143. Students who fail or withdraw from this course multiple times may be prohibited from majoring in the sciences or mathematics; seethe bulletin language for your major

### Professor: Steven Goldman, AVG GPA: 2.9 (https://www.ratemyprofessors.com/professor/1786967) which is a B-

## Math 209 (Elementary Set Theory) - Basic axioms of set theory, algebra of sets, relations and functions, orders, countable and uncountable sets, and additional topics at the discretion of the instructor. The course will introduce some basic proof techniques, with no background in proof-writing assumed. Not open to students who are taking or have received credit for MATH 509 or 609 

### Professor: Krzysztof Klosin, AVG GPA: 2.2 (https://www.ratemyprofessors.com/professor/1650543) which is a C

## Math 242 (Methods of Math Statistics) - A study of those methods of mathematical statistics that are most frequently used in the natural and social sciences, as well as actuarial science. Topics include estimation testing of statistical hypotheses, nonparametric tests, analysis of variance, correlation and regression analysis, and other methods of statistical analysis

### Professor: Fern Sisser, AVG GPA: 2.2 (https://www.ratemyprofessors.com/professor/22553) which is a C

## Math 247 (Lin Prog and Game Theory) - Methods for handling optimization problems that arise in management, engineering, physical sciences, and social sciences. Topics include convex geometry, the simplex algorithm, duality theory, and the Von Neumann minimax theorem of game theory 

### Professor: Krzysztof Klosin, AVG GPA: 3.3 (https://www.ratemyprofessors.com/professor/1650543) which is a A-

## Math 301 (Abstract Algebra I) - Theory of groups, including cyclic and permutation groups, homomorphisms, normal subgroups and quotient groups. Theory of rings, including integral domains and polynomial rings. Additional topics may be discussed. Not open to students who are taking or who have received credit for MATH 601 or 702

### Professor: Russel Miller, AVG GPA: 0.8, everyone dropped or failed (https://www.ratemyprofessors.com/professor/311326) which is an F

## Math 310 (Elementary Real Analysis) - Rigorous introduction to functions of a real variable. Topics include real numbers and the completeness property; limits of sequences; elementary topological concepts; continuity and uniform continuity; sequences and series of functions, derivatives; Taylor’s theorem; the Riemann integral 

### Professor: Scott Wilson, AVG GPA: 3.1 (https://www.ratemyprofessors.com/professor/1157817) which is a B+

## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

#### Taking all the one off classes I analyzed and removing them from the main CS/Math dataframe 

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html
https://stackoverflow.com/questions/37313691/how-to-remove-a-pandas-dataframe-from-another-dataframe

In [18]:
pd.concat([CS_MATH_df, one_class_df]).drop_duplicates(keep = False)

SUBJECT     NBR                     COURSE NAME               PROF  \
585     CSCI    48.0         Spreadsheet Programming           FRIED, M   
586     CSCI    48.0         Spreadsheet Programming          JAGDEO, M   
597     CSCI   111.0  Intro Algorithmic Problem Solv          SCHLEY, R   
598     CSCI   111.0  Intro Algorithmic Problem Solv            CHYN, X   
599     CSCI   111.0  Intro Algorithmic Problem Solv            CHYN, X   
621     CSCI   211.0  Object-Oriented Program in C++          WAXMAN, J   
622     CSCI   211.0  Object-Oriented Program in C++          WAXMAN, J   
623     CSCI   211.0  Object-Oriented Program in C++          ALAYEV, Y   
635     CSCI   212.0    Object-Oriented Prog in Java             TSE, C   
636     CSCI   212.0    Object-Oriented Prog in Java            LORD, K   
637     CSCI   212.0    Object-Oriented Prog in Java            LORD, K   
648     CSCI   220.0             Discrete Structures       KAHROBAEI, D   
649     CSCI   220.0             Discrete Structures       KAHROBAEI, D   
650     CSCI   220.0             Discrete Structures            CHYN, X   
651     CSCI   220.0             Discrete Structures            CHYN, X   
652     CSCI   220.0             Discrete Structures       CESARETTI, P   
653     CSCI   240.0    Computer Org & Assembly Lang             YEH, J   
654     CSCI   240.0    Computer Org & Assembly Lang             YEH, J   
655     CSCI   240.0    Computer Org & Assembly Lang             YEH, J   
656     CSCI   240.0    Computer Org & Assembly Lang          SVITAK, J   
657     CSCI   240.0    Computer Org & Assembly Lang          SVITAK, J   
658     CSCI   240.0    Computer Org & Assembly Lang          SVITAK, J   
659     CSCI   313.0                 Data Structures             TAO, X   
660     CSCI   313.0                 Data Structures             TAO, X   
661     CSCI   313.0                 Data Structures             TAO, X   
662     CSCI   313.0                 Data Structures            TSAI, C   
663     CSCI   313.0                 Data Structures       STEINBERG, O   
664     CSCI   313.0                 Data Structures  SMITH-THOMPSON, A   
665     CSCI   313.0                 Data Structures       STEINBERG, O   
666     CSCI   313.0                 Data Structures  SMITH-THOMPSON, A   
667     CSCI   316.0  Principles of Programming Lang  SMITH-THOMPSON, A   
668     CSCI   316.0  Principles of Programming Lang            KONG, T   
669     CSCI   316.0  Principles of Programming Lang            KONG, T   
670     CSCI   316.0  Principles of Programming Lang          SVITAK, J   
671     CSCI   316.0  Principles of Programming Lang          SVITAK, J   
672     CSCI   316.0  Principles of Programming Lang            KONG, T   
673     CSCI   320.0           Theory of Computation         OBRENIC, B   
674     CSCI   320.0           Theory of Computation          BOKLAN, K   
675     CSCI   320.0           Theory of Computation         OBRENIC, B   
676     CSCI   320.0           Theory of Computation          BOKLAN, K   
677     CSCI   323.0    Design & Analysis Algorithms             TAO, X   
678     CSCI   323.0    Design & Analysis Algorithms        PHILLIPS, T   
679     CSCI   323.0    Design & Analysis Algorithms        PHILLIPS, T   
680     CSCI   323.0    Design & Analysis Algorithms           BROWN, T   
681     CSCI   323.0    Design & Analysis Algorithms         GOSWAMI, M   
682     CSCI   331.0                Database Systems       STEINBERG, O   
683     CSCI   331.0                Database Systems       STEINBERG, O   
684     CSCI   331.0                Database Systems          HELLER, P   
685     CSCI   331.0                Database Systems         LEAVITT, D   
686     CSCI   331.0                Database Systems         OBRENIC, B   
689     CSCI   340.0    Operating Systems Principles       SVADLENKA, J   
690     CSCI   340.0    Operating Systems Principles       SVADLENKA, J   
691     CSCI   340.0    Operat

In [19]:
CS_MATH_df = pd.concat([CS_MATH_df, one_class_df]).drop_duplicates(keep = False)
CS_MATH_df[CS_MATH_df["NBR"] <= 399]

SUBJECT    NBR                     COURSE NAME               PROF  TOTAL  \
585     CSCI   48.0         Spreadsheet Programming           FRIED, M    131   
586     CSCI   48.0         Spreadsheet Programming          JAGDEO, M     94   
597     CSCI  111.0  Intro Algorithmic Problem Solv          SCHLEY, R    236   
598     CSCI  111.0  Intro Algorithmic Problem Solv            CHYN, X    239   
599     CSCI  111.0  Intro Algorithmic Problem Solv            CHYN, X    240   
621     CSCI  211.0  Object-Oriented Program in C++          WAXMAN, J    150   
622     CSCI  211.0  Object-Oriented Program in C++          WAXMAN, J    105   
623     CSCI  211.0  Object-Oriented Program in C++          ALAYEV, Y     16   
635     CSCI  212.0    Object-Oriented Prog in Java             TSE, C     36   
636     CSCI  212.0    Object-Oriented Prog in Java            LORD, K    161   
637     CSCI  212.0    Object-Oriented Prog in Java            LORD, K    127   
648     CSCI  220.0             Discrete Structures       KAHROBAEI, D     39   
649     CSCI  220.0             Discrete Structures       KAHROBAEI, D     42   
650     CSCI  220.0             Discrete Structures            CHYN, X     88   
651     CSCI  220.0             Discrete Structures            CHYN, X     58   
652     CSCI  220.0             Discrete Structures       CESARETTI, P     30   
653     CSCI  240.0    Computer Org & Assembly Lang             YEH, J     45   
654     CSCI  240.0    Computer Org & Assembly Lang             YEH, J     45   
655     CSCI  240.0    Computer Org & Assembly Lang             YEH, J     45   
656     CSCI  240.0    Computer Org & Assembly Lang          SVITAK, J     31   
657     CSCI  240.0    Computer Org & Assembly Lang          SVITAK, J     26   
658     CSCI  240.0    Computer Org & Assembly Lang          SVITAK, J     37   
659     CSCI  313.0                 Data Structures             TAO, X     39   
660     CSCI  313.0                 Data Structures             TAO, X     36   
661     CSCI  313.0                 Data Structures             TAO, X     24   
662     CSCI  313.0                 Data Structures            TSAI, C     10   
663     CSCI  313.0                 Data Structures       STEINBERG, O     31   
664     CSCI  313.0                 Data Structures  SMITH-THOMPSON, A     29   
665     CSCI  313.0                 Data Structures       STEINBERG, O     42   
666     CSCI  313.0                 Data Structures  SMITH-THOMPSON, A     25   
667     CSCI  316.0  Principles of Programming Lang  SMITH-THOMPSON, A     35   
668     CSCI  316.0  Principles of Programming Lang            KONG, T     35   
669     CSCI  316.0  Principles of Programming Lang            KONG, T     35   
670     CSCI  316.0  Principles of Programming Lang          SVITAK, J     28   
671     CSCI  316.0  Principles of Programming Lang          SVITAK, J     30   
672     CSCI  316.0  Principles of Programming Lang            KONG, T     61   
673     CSCI  320.0           Theory of Computation         OBRENIC, B     86   
674     CSCI  320.0           Theory of Computation          BOKLAN, K     37   
675     CSCI  320.0           Theory of Computation         OBRENIC, B    109   
676     CSCI  320.0           Theory of Computation          BOKLAN, K     28   
677     CSCI  323.0    Design & Analysis Algorithms             TAO, X     30   
678     CSCI  323.0    Design & Analysis Algorithms        PHILLIPS, T     18   
679     CSCI  323.0    Design & Analysis Algorithms        PHILLIPS, T     25   
680     CSCI  323.0    Design & Analysis Algorithms           BROWN, T     30   
681     CSCI  323.0    Design & Analysis Algorithms         GOSWAMI, M     61   
682     CSCI  331.0                Database Systems       STEINBERG, O     35   
683     CSCI  331.0                Database Systems       STEINBERG, O     31   
684     CSCI  331.0                Database Systems          HELLER, P     29   
685     CSCI  331.0                D